### PR TITLE
doc: adding a note about config expressions

### LINKED ISSUE #19831: Special characters in `--db-password` option
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

core

### Describe the bug

If i'm trying with some other application to same database i'm able to connect to SQL server , infact if i use the Driver to test connection i'm able to do that.

But when I'm using running Keycloak it fails to connect and throws the error.

2023-04-19 14:59:37,713 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to obtain JDBC connection
2023-04-19 14:59:37,713 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Login failed for user 'keycloak-user'. ClientConnectionId:b3662e81-673b-4b0e-a2fc-906f8548e325
2023-04-19 14:59:37,713 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option.

### Version

20.0.3

### Expected behavior

It should be able to connect 

### Actual behavior

2023-04-19 14:59:37,713 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to obtain JDBC connection
2023-04-19 14:59:37,713 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Login failed for user 'keycloak-user'. ClientConnectionId:b3662e81-673b-4b0e-a2fc-906f8548e325
2023-04-19 14:59:37,713 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option.

### How to Reproduce?

Try to connect sql server with named instance which is running in remote machine.

### Anything else?

Please fix asap , I'm blocked with it.

### PR DESCRIPTION
**Pull request by shawkins**
I'm not sure if there's even more nuance to express here as some locations we may specifically look at the raw values instead.

There hasn't been any follow up on https://github.com/smallrye/smallrye-config/issues/1056 to know if the $$ behavior is expected.

fix: #19831

### CODE REVIEW

File: docs/guides/server/configuration.adoc [hunk 1]
```diff
@@ -143,10 +143,16 @@ You can use only a https://github.com/keycloak/keycloak/blob/main/quarkus/runtim
...
-Similarly, you can also store Quarkus properies in a Java KeyStore.
+Similarly, you can also store Quarkus properties in a Java KeyStore.
 
 Note that some Quarkus properties are already mapped in the {project_name} configuration, such as `quarkus.http.port` and similar essential properties. If the property is used by Keycloak, defining that property key in `quarkus.properties` has no effect. The Keycloak configuration value takes precedence over the Quarkus property value.
 
+== Values
+
+Keycloak depends upon Quarkus / MicroProfile for processing configuration values. Be aware this means that value expression are supported, for example value=$\{some_value}$\{some_other_value}, which will evaluate to something other than the literal value.
+
+To disable expression evaluation, the \ character is used as an escape character where needed. In particular it must be used to escape the usage of $ characters when they appear to define an expression or are repeated.  For example instead of a configuration value of my$$password, use my\$\$password.
```
[the remark this answers is not in the document]
Comment (ahus1) [reply docs/guides/server/configuration.adoc:154]: Some comments here around quoting contents for the docs (apply here if necessary):

* If you struggle with escaping things for AsciiDoc, consider putting it into double-plus signs for pass-through content (`++passthrough++`).
* If you struggle with escaping freemarker stuff (like things in curly braces), wrap it as no-parse (`<#noparse>...</#noparse>`)

[the remark this answers is not in the document]
Comment (shawkins) [reply docs/guides/server/configuration.adoc:154]: > Seems triple \$ is needed?

No, the key here is using \\ to escape the $.  Using \\$\\$\\$ would evaluate to 3 dollar signs, not 2.  There may be a fix upstream to correct the $$ behavior, so we don't want to document relying on that - https://github.com/smallrye/smallrye-config/issues/1056

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/server/configuration.adoc:154]: @shawkins Not sure why (haven't investigated) but if I use something like `--db-username posgre\$\$\$ql`, it evaluates to `posgre$$ql` (looking at DB log of failed auth attempt).

[the remark this answers is not in the document]
Comment (shawkins) [reply docs/guides/server/configuration.adoc]: @vmuzikar please review again, I've incorporated most the suggestions and further clarified shell handling.

[the remark this answers is not in the document]
Comment (vmuzikar) [reply docs/guides/server/configuration.adoc]: I looked into it more. The issue with 3 `$$$` was truly caused by the shell backslash escaping as you hinted.

However, I think the docs change is a little bit confusing still. We're mixing together two things: the SmallRye expression escaping, and shell expressions escaping. For example, let's assume the `my$$password` value. When specified in shell, it needs to be entered as `"my\$\$password"` to mitigate for shell expressions. However, when specified via a properties file it needs to be entered as `my$$$$password` to mitigate SmallRye expressions (which are not evaluated on other config sources AFAIK). This is currently not captured in the docs.

[the remark this answers is not in the document]
Comment (shawkins) [reply docs/guides/server/configuration.adoc]: > We're mixing together two things: the SmallRye expression escaping, and shell expressions escaping.

That's not intended - shell escaping only comes into this at the very end as an additional concern on top of the escaping needed for Microprofile config / smallrye.

> For example, let's assume the my$$password value. When specified in shell, it needs to be entered as "my\$\$password" to mitigate for shell expressions.

That's not at all what I'm trying to convey - it doesn't matter where the value appears, the $$ needs to be escaped.

> However, when specified via a properties file it needs to be entered as my$$$$password to mitigate SmallRye expressions

That's not at all what I would recommend.  As far as I can find, that's undocumented behavior - unless I'm missing something it doesn't appear in Microprofile config nor quarkus docs.  I do see references to \ as the escape character - and it's explicitly handled in the smallrye code as such wrt $.

> which are not evaluated on other config sources AFAIK

The removal of $$ is certainly happening when processing cli args, are you suggesting that you don't expect nested expressions to be resolved with cli or env variables?

**Comment by shawkins** [comment]
@vmuzikar see if this further clarifies things.  It doesn't appear that additional general \ escape handling is coming from upstream anytime soon, so I wouldn't explicitly mention anything in our docs about that.  We can of course also push for quarkus to take ownership of spelling out escaping rules as well.

File: docs/guides/server/configuration.adoc [hunk 2]
```diff
@@ -143,10 +143,17 @@ You can use only a https://github.com/keycloak/keycloak/blob/main/quarkus/runtim
...
 
+=== Using special characters in values
+
+Keycloak depends upon Quarkus / MicroProfile for processing configuration values. Be aware that value expressions are supported - for example `$\{some_key}`, will evaluate to the value of some_key.
+
+To disable expression evaluation, the `\` character is used as an escape character. In particular it must be used to escape the usage of `$` characters when they appear to define a expression or are repeated. For example if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.  
+For example bash single quotes will preseve the single backslash --db-password=`'my\$\$password'`, or with bash double quotes you will need an additional backslash --db-password=`"my\\$\\$password"`. Similarly in a properties file backslash characters must themselves be escaped: kc.db-password=`my\\$\\$password`
```
Comment (vmuzikar) [inline docs/guides/server/configuration.adoc:155]: Thanks for the update! I think it's clearer now.

One last nitpick. I think we should be putting whole key+value in the backticks. Seems It's more common throughout the docs.
```suggestion
To disable expression evaluation, the `\` character is used as an escape character. In particular it must be used to escape the usage of `$` characters when they appear to define a expression or are repeated. For example if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.  
For example bash single quotes will preserve the single backslash `--db-password='my\$\$password'`, or with bash double quotes you will need an additional backslash `--db-password="my\\$\\$password"`. Similarly in a properties file backslash characters must themselves be escaped: `kc.db-password=my\\$\\$password`
```

**Review by vmuzikar — DISMISSED** [review]

**Comment by shawkins** [comment]
@ahus1 do you want to review this before it is merged?

**Comment by ahus1** [comment]
@shawkins - I pushed one more commit to fix a typo (`a` -> `an`) and added a replacement for the project name with we need to do always. The single `/` is also very uncommon in our docs. 

For further writing style as some new paragraphs are added, I suggest to give @andymunro a chance to review the docs from a technical writing perspective.

**Review by ahus1 — COMMENTED** [review]
Once Andy's review is in, this should can be squashed and merged.

File: docs/guides/server/configuration.adoc [hunk 3]
```diff
@@ -143,10 +143,17 @@ You can use only a https://github.com/keycloak/keycloak/blob/main/quarkus/runtim
...
 . Use the `[-cf|--config-file]` command line parameter to include that file.
 
-Similarly, you can also store Quarkus properies in a Java KeyStore.
+Similarly, you can also store Quarkus properties in a Java KeyStore.
 
 Note that some Quarkus properties are already mapped in the {project_name} configuration, such as `quarkus.http.port` and similar essential properties. If the property is used by Keycloak, defining that property key in `quarkus.properties` has no effect. The Keycloak configuration value takes precedence over the Quarkus property value.
 
+=== Using special characters in values
+
+{project_name} depends upon Quarkus and MicroProfile for processing configuration values. Be aware that value expressions are supported - for example `$\{some_key}`, will evaluate to the value of `some_key`.
```
Comment (andymunro) [inline docs/guides/server/configuration.adoc:152]:
```suggestion
{project_name} depends upon Quarkus and MicroProfile for processing configuration values. Be aware that value expressions are supported. For example, `$\{some_key}` evaluates to the value of `some_key`.
```

File: docs/guides/server/configuration.adoc [hunk 4]
```diff
@@ -143,10 +143,17 @@ You can use only a https://github.com/keycloak/keycloak/blob/main/quarkus/runtim
...
+
+To disable expression evaluation, the `\` character is used as an escape character. In particular it must be used to escape the usage of `$` characters when they appear to define an expression or are repeated. For example if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.
```
Comment (andymunro) [inline docs/guides/server/configuration.adoc:154]:
```suggestion
To disable expression evaluation, the `\` character functions as an escape character. In particular, it must be used to escape the usage of `$` characters when they appear to define an expression or are repeated. For example, if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.
```

File: docs/guides/server/configuration.adoc [hunk 5]
```diff
@@ -143,10 +143,17 @@ You can use only a https://github.com/keycloak/keycloak/blob/main/quarkus/runtim
...
+For example bash single quotes will preserve the single backslash `--db-password='my\$\$password'`, or with bash double quotes you will need an additional backslash `--db-password="my\\$\\$password"`. Similarly in a properties file backslash characters must themselves be escaped: `kc.db-password=my\\$\\$password`
```
Comment (andymunro) [inline docs/guides/server/configuration.adoc:155]:
```suggestion
For example, bash single quotes preserve the single backslash `--db-password='my\$\$password'`. Also, with bash double quotes, you need an additional backslash `--db-password="my\\$\\$password"`. Similarly in a properties file, backslash characters must also be escaped: `kc.db-password=my\\$\\$password`
```

**Review by andymunro — CHANGES_REQUESTED** [review]
Just a few changes.

**Review by ahus1 — APPROVED** [review]
Approving as of @vmuzikar's previous review and as all of @andymunro's changes have been accepted. Ready to be merged!

**Comment by shawkins** [comment]
@ahus1 @andymunro thank you for the updates, they've been squashed into the review commit.

**Pull request merged**